### ISSUE TITLE
关于模型的一些问题

### ISSUE THREAD

**Issue by nlper01** [ljynlp/W2NER#69]
大佬，又有几个新的疑问：
1、模型为什么最后能学到上下三角不一样的 Word-Word Relation Classification 矩阵？
（1）前面的Word Embedding【B，L，L，512】应该是一个对称三角吧，为什么最后学到一个上下三角不一样的？如果说是因为前面的 Region Embedding 有上下三角的区别，Distance Embedding 上下三角距离也做了类似的区分，可让 Region Embedding 不再区分上下三角，把 Distance Embedding 去掉对结果影响也不是特别大，说明不是因为这两个 Embedding？

（2）前面的Word Embedding【B，L，L，Embeds】应该是一个对称三角，和Region Embedding ，Distance Embedding进行concat后【B，L，L，552】有了上下三角区分，但是经过后面的卷积——>【B，L，L，228】感觉把这个差异抹除了，再经过MLP线性变换——>【B，L，L，C】还能保证上下三角差异吗？
_**这里有点疑惑。**_

2、Word-Word Relation Classification 会不会出现学乱的情况，比如：上三角最大值索引是（3，4），下三角最大值索引是（0，1）这种情况？

**_期待大佬解惑，感谢感谢！_**

**Comment by ljynlp** [comment]
1. Word Embedding并不是对称的，因为 $V_{ij}=CLN(h_i, cond=h_j)$ ，而 $V_{ji}=CLN(h_j, cond=h_i)$ ，两者并不等价。
2. 在此基础上我们继续加入Region信息和Distance信息，卷积并不会抹除差异，而是用于学习2d平面上的word之间的交互与临近信息。
3. 学乱的情况没有理解你举的例子。

**Comment by nlper01** [comment]
学乱的情况，我的意思是：Co-pre 的输出经过 torch.argmax ( ) 得到Word-Word Relation Classification ，这个Word-Word Relation Classification应该是上三角区域只有0,1两种值，下三角是2,3,4.......等实体类型，那有没有可能这里出现上三角出现了3,4，下三角出现0,1这种情况？

**Comment by ljynlp** [comment]
> 学乱的情况，我的意思是：Co-pre 的输出经过 torch.argmax ( ) 得到Word-Word Relation Classification ，这个Word-Word Relation Classification应该是上三角区域只有0,1两种值，下三角是2,3,4.......等实体类型，那有没有可能这里出现上三角出现了3,4，下三角出现0,1这种情况？

0表示没有关系，严格意义上来说上三角有0,1，下三角有0,2,3,4......，在解码的时候我们只对上三角解码0,1，其他关系直接忽略，下三角也同理

**Comment by nlper01** [comment]
懂了，谢谢大佬

**Comment by nlper01** [comment]
大佬，不好意思，又有个地方要请教一下。卷积层输出是【L，L，288】，然后经过一个MLP线性变换【L，L，label_num】，双仿射分类器也是同理，最后再在label_num这一维上用torch.agmax(-1)得到关系预测矩阵【L，L】，这里有点懵，为什么最后一维能表示关系类型呢？是288那一维学到的是关于词对关系的信息吗？

**Comment by ljynlp** [comment]
最后的维度由label_num变成1是因为agmax()取了label_num里面概率最大的那个label的索引，所以关系预测矩阵[L，L]中每个元素都是一个关系label的索引，因此可以解码出词对关系。

**Comment by nlper01** [comment]
> 最后的维度由label_num变成1是因为agmax()取了label_num里面概率最大的那个label的索引，所以关系预测矩阵[L，L]中每个元素都是一个关系label的索引，因此可以解码出词对关系。 这个我理解了。
我想问的是：【L，L，288】——>【L，L，label_num】这里为什么在最后一维上变换呢？是288那一维学到的是关于词对关系的信息吗？

**Comment by ljynlp** [comment]
是的

**Comment by FalAnge1217** [comment]
非常感谢大佬耐心解答！这里模板了，还想问个小白问题，损失函数这里。
1、这种建模方式是属于多分类吧？还是多标签分类？
2、Co-Predictor Layer 出来经过softmax激活，然后再计算损失，代码里用了CrossEntropyLoss，具体怎么计算的有点没搞明白。
           ` loss = self.criterion(outputs[grid_mask2d], grid_labels[grid_mask2d])  `
比如对于这个样本{`"sentence": ["高", "勇", "：", "男", "."], "ner": [{"index": [0, 1], "type": "NAME"}]}`
最后得到的`outputs[grid_mask2d]`（input）和`grid_labels[grid_mask2d]`（target）如下：
![图片](https://user-images.githubusercontent.com/82627285/194707653-bc81b5a2-636f-4e5a-b9a6-255dd63886f2.png)
这里具体是怎么计算的，没搞明白，能否举个示例。

**Comment by nlper01** [comment]
> 学乱的情况，我的意思是：Co-pre 的输出经过 torch.argmax ( ) 得到Word-Word Relation Classification ，这个Word-Word Relation Classification应该是上三角区域只有0,1两种值，下三角是2,3,4.......等实体类型，那有没有可能这里出现上三角出现了3,4，下三角出现0,1这种情况？
> 
> 0表示没有关系，严格意义上来说上三角有0,1，下三角有0,2,3,4......，在解码的时候我们只对上三角解码0,1，其他关系直接忽略，下三角也同理

大佬，这种标注方式实体和类型标签是分开标注的，那这样不可避免的就会出现类型和边界的对应不上的问题，比如边界预测对了类型没有预测对，或者反过来，这种情况怎么处理的呢？

**Comment by ljynlp** [comment]
你可以看看解码策略，预测错误就可能导致解码不出来或者解出错误的实体，最终也会影响召回率和精准率，在指标上也有体现
